### ISSUE TITLE
Issue better errors for `?.` and `!.`in type-space

### ISSUE THREAD

**Issue by weswigham** [microsoft/TypeScript#57244]
### 🔎 Search Terms

typeof optional chain type error

### 🕗 Version & Regression Information

current

### 💻 Code

```ts
type Q1 = A?.B
type Q2 = A!.B
type Q3 = typeof A?.B
type Q4 = typeof A!.B
```

### 🙁 Actual behavior

Bad error.

### 🙂 Expected behavior

`Optional chains are not allowed in entity names.` or `Non-null assertions are not allowed in entity names.`

### Additional information about the issue

Productions like `type Q = A?.B` or `type Q = A!.B` produce a pretty bad `, expected` type error today. This is probably a common error - we should improve it.

**Comment by weswigham** [comment]
Do note, this shouldn't preclude actually supporting `?.` as a nice shorthand in types, same as it is in expressions. That's basically tracked by #45793, which we really should also consider.

**Comment by fatcerberus** [comment]
What would an optional chain in typespace even *mean*?  My best guess is that `A?.B` would evaluate to `A["B"] | undefined` (ignoring `null` and `undefined` if they're part of the type of `A`).

**Comment by DanielRosenwasser** [comment]
Right, something like `NonNullable<A>["B"] | undefined` (though losing information about whether `A` contained `null` is questionable - but it always sort of was in JS too).
